src/ptvsd/adapter/ide.py

```diff
@@ -222,58 +260,101 @@ def launch_request(self, request):
 
-            host = request("host", "127.0.0.1")
-            port = request("port", int)
-            if request("listen", False):
-                with self.accept_connection_from_server((host, port)):
-                    pass
-            else:
-                self.session.connect_to_server((host, port))
-        else:
-            if self.server:
+        pid = request("processId", int, optional=True)
+        sub_pid = request("subProcessId", int, optional=True)
+        if pid != ():
+            if sub_pid != ():
                 raise request.isnt_valid(
-                    '"attach" with "processId" cannot be serviced by adapter '
-                    "that is already associated with a debug server"
+                    '"processId" and "subProcessId" are mutually exclusive'
                 )
-
             ptvsd_args = request("ptvsdArgs", json.array(unicode))
-            self.session.inject_server(pid, ptvsd_args)
+            server.inject(pid, ptvsd_args)
+            timeout = 10
+        else:
+            if sub_pid == ():
+                pid = any
+                timeout = None if request("waitForAttach", False) else 0
+            else:
+                pid = sub_pid
+                timeout = 0
+
+        conn = server.wait_for_connection(pid, timeout)
+        if conn is None:
+            raise request.cant_handle(
+                (
+                    "Timed out waiting for injected debug server to connect"
+                    if timeout
+                    else "There is no debug server connected to this adapter."
+                    if pid is any
+                    else 'No known subprocess with "subProcessId":{0}'
+                ),
+                pid,
+            )
+
+        try:
+            conn.attach_to_session(self.session)
+        except ValueError:
+            request.cant_handle("Debuggee with PID={0} is already being debugged.", pid)
 
     @message_handler
     def configurationDone_request(self, request):
-        if self._start_request is None:
+        if self.start_request is None or self.has_started:
             request.cant_handle(
                 '"configurationDone" is only allowed during handling of a "launch" '
                 'or an "attach" request'
             )
 
         try:
+            self.has_started = True
             request.respond(self.server.channel.delegate(request))
+        except messaging.MessageHandlingError as exc:
+            self.start_request.cant_handle(str(exc))
         finally:
-            self._start_request.respond({})
-            self._start_request = None
+            self.start_request.respond({})
             self._propagate_deferred_events()
 
+        # Notify the IDE of any child processes of the debuggee that aren't already
+        # being debugged.
+        for conn in server.connections():
+            if conn.server is None and conn.ppid == self.session.pid:
+                # FIXME: race condition with server.Connection()
```

Comment:
Do we need a bug for this?

Reply:
Yep. I'll file one after merging the PR, so that I can reference the commit/line directly.